### PR TITLE
Internal errors parsing response are not triggering "catch" functions

### PR DESCRIPTION
**Pull request by Kibet-mutai**
In the` json` function, the `JsonDecodeException` is caught, but instead of passing it to the catch function in the promise chain, it is being handled internally by rejecting the new promise created with `$newDeferred->reject($exception)`;. This means that the rejection is not being propagated through the promise chain to the `catch `function, which is why the exception is not being caught.

### CODE REVIEW

**Comment by Kibet-mutai** [comment]
On my also it didn't `echo` exception but I think it did catch the exception, I will look where I went wrong.

**Comment by g105b** [comment]
Yes, please see the comment in the [original issue's code, near the bottom](https://github.com/PhpGt/Fetch/issues/103):

> `// THIS FUNCTION SHOULD TRIGGER, BUT DOESN'T!`

**Comment by Kibet-mutai** [comment]
Is the message: `"Error decoding JSON: Syntax error"` type: `Gt\Json\JsonDecodeException`?

**Comment by g105b** [comment]
In this case, yes it will be. But there may be other reasons the JSON parsing fails, so that's why we're using a `Throwable` which could represent any sort of exception or error.

Take a look at this code I just wrote: https://codepen.io/g105b/pen/LYgpayW - it's a JavaScript version of the example code we're using, to show you how the browser works with the Fetch API.

**Comment by Kibet-mutai** [comment]
So I reproduce another Json file with other error other than syntax error in the json file?

**Comment by g105b** [comment]
You can use the broken JSON we've been using. You don't need to worry about other errors, I was just being thorough.

File: src/Response/BodyResponse.php [hunk 1]
```diff
@@ -127,6 +128,8 @@ public function json(int $depth = 512, int $options = 0): Promise
 			} catch (JsonDecodeException $exception) {
 				$newDeferred->reject($exception);
 			}
+		})->catch(function (Throwable $reason) use ($newDeferred) {
```
Comment (Kibet-mutai) [inline src/Response/BodyResponse.php:131]: image
I added catch function at the end here in the srch/Response/BodyResponse.php

**Comment by Kibet-mutai** [comment]
> You can use the broken JSON we've been using. You don't need to worry about other errors, I was just being thorough.

Okay

**Comment by g105b** [comment]
> I think the problem is with `return $response->json()` in the example [but with inavlid Json](https://github.com/PhpGt/Fetch/issues/103) it doesn't raise an exception when given invalid Json data.

As you found [in your previous message](https://github.com/PhpGt/Fetch/pull/137#pullrequestreview-1381969461), the exception is being triggered correctly _somewhere_, but you're right - the exception isn't making its way to the `catch()` function of the promise chain.

One thing to note is that `return $response->json()` is a line of code that can't change, because that's how the specification states a JSON response should be handled.

So if there's a `catch()` function in the promise chain, the exception should not _throw_ - instead, the catch() function should be executed and passed the appropriate exception as a parameter.

**Comment by Kibet-mutai** [comment]
> > I think the problem is with `return $response->json()` in the example [but with inavlid Json](https://github.com/PhpGt/Fetch/issues/103) it doesn't raise an exception when given invalid Json data.
> 
> As you found [in your previous message](https://github.com/PhpGt/Fetch/pull/137#pullrequestreview-1381969461), the exception is being triggered correctly _somewhere_, but you're right - the exception isn't making its way to the `catch()` function of the promise chain.
> 
> One thing to note is that `return $response->json()` is a line of code that can't change, because that's how the specification states a JSON response should be handled.
> 
> So if there's a `catch()` function in the promise chain, the exception should not _throw_ - instead, the catch() function should be executed and passed the appropriate exception as a parameter.

I commented the return `$response->json()` and `catch()` function echoed out an error. Also I introduce other error in the basic 01-example and exception throws and error. 

Here is error message in the debug console: `ERROR: {closure}(): Argument #1 ($json) must be of type Gt\Json\JsonPrimitive\JsonArrayPrimitive, Gt\Fetch\Response\BodyResponse given`
So you mean there is no way we can change the `return $reponse->json()` ?

**Comment by Kibet-mutai** [comment]
I remove the `catch() `function  in the 01-basic-Fetch example, I used an invalid Json we've been using, instead I took the line `echo "ERROR: ", $reason->getMessage(), PHP_EOL, PHP_EOL;` and I added in the [BodyResponse.php](https://github.com/PhpGt/Fetch/blob/aea13da44f83f08e650badbaf06f42dba51ca581/src/Response/BodyResponse.php#L132) and I got the error echoed as expected when the exception is caught. 
As in instead of `catch()` function in the example we should place in the [json function](https://github.com/PhpGt/Fetch/blob/aea13da44f83f08e650badbaf06f42dba51ca581/src/Response/BodyResponse.php#L117).
That's my suggestion, reach out if any question.

**Comment by g105b** [comment]
The reason we want to see `ERROR: Error decoding JSON: Syntax error` is not just so we can echo the error message, but because it means the `catch()` function is being executed correctly, according to the [fetch standard](https://fetch.spec.whatwg.org/#fetch-method).

If we removed the `catch()` and catch the exception manually in the `BodyResponse`, the library is not built according to the standard.

**Comment by Kibet-mutai** [comment]
I have been working on the bug, however, in the `Gt\Fetch\Http`, it seems that the JSON decoding is not handled properly, which causes the uncaught exception.
Does, `Gt\Fetch\Http` class has function which handles JSON DECODE EXCEPTION?

**Comment by g105b** [comment]
This is now resolved in the latest commit.

**Pull request closed without merging**